sys/linux/fs_ioctl_btrfs.txt

```diff
@@ -1,7 +1,426 @@
+#ioctl$BTRFS_IOC_SEND_32(fd fd, cmd const[BTRFS_IOC_SEND_32], arg ptr[in, btrfs_ioctl_send_args])
+ioctl$BTRFS_IOC_GET_DEV_STATS(fd fd, cmd const[BTRFS_IOC_GET_DEV_STATS], arg ptr[inout, btrfs_ioctl_get_dev_stats])
+ioctl$BTRFS_IOC_QUOTA_CTL(fd fd, cmd const[BTRFS_IOC_QUOTA_CTL], arg ptr[in, btrfs_ioctl_quota_ctl_args])
+ioctl$BTRFS_IOC_QGROUP_ASSIGN(fd fd, cmd const[BTRFS_IOC_QGROUP_ASSIGN], arg ptr[in, btrfs_ioctl_qgroup_assign_args])
+ioctl$BTRFS_IOC_QGROUP_CREATE(fd fd, cmd const[BTRFS_IOC_QGROUP_CREATE], arg ptr[in, btrfs_ioctl_qgroup_create_args])
+ioctl$BTRFS_IOC_QGROUP_LIMIT(fd fd, cmd const[BTRFS_IOC_QGROUP_LIMIT], arg ptr[in, btrfs_ioctl_qgroup_limit_args])
+ioctl$BTRFS_IOC_QUOTA_RESCAN(fd fd, cmd const[BTRFS_IOC_QUOTA_RESCAN], arg ptr[in, btrfs_ioctl_quota_rescan_args])
+ioctl$BTRFS_IOC_QUOTA_RESCAN_STATUS(fd fd, cmd const[BTRFS_IOC_QUOTA_RESCAN_STATUS], arg ptr[out, btrfs_ioctl_quota_rescan_args])
+ioctl$BTRFS_IOC_QUOTA_RESCAN_WAIT(fd fd, cmd const[BTRFS_IOC_QUOTA_RESCAN_WAIT], arg const[0])
+ioctl$BTRFS_IOC_DEV_REPLACE(fd fd, cmd const[BTRFS_IOC_DEV_REPLACE], arg ptr[inout, btrfs_ioctl_dev_replace_args])
+
+#the arg should be of length 3. Not sure how to specify that
+ioctl$BTRFS_IOC_GET_SUPPORTED_FEATURES(fd fd, cmd const[BTRFS_IOC_GET_SUPPORTED_FEATURES], arg ptr[out, btrfs_ioctl_feature_flags])
+
+ioctl$BTRFS_IOC_GET_FEATURES(fd fd, cmd const[BTRFS_IOC_GET_FEATURES], arg ptr[out, btrfs_ioctl_feature_flags])
+ioctl$BTRFS_IOC_SET_FEATURES(fd fd, cmd const[BTRFS_IOC_SET_FEATURES], arg ptr[in, btrfs_ioctl_feature_flags])
+ioctl$BTRFS_IOC_GET_SUBVOL_INFO(fd fd, cmd const[BTRFS_IOC_GET_SUBVOL_INFO], arg ptr[out, btrfs_ioctl_get_subvol_info_args])
+ioctl$BTRFS_IOC_GET_SUBVOL_ROOTREF(fd fd, cmd const[BTRFS_IOC_GET_SUBVOL_ROOTREF], arg ptr[inout, btrfs_ioctl_get_subvol_rootref_args])
+ioctl$BTRFS_IOC_INO_LOOKUP_USER(fd fd, cmd const[BTRFS_IOC_INO_LOOKUP_USER], arg ptr[inout, btrfs_ioctl_ino_lookup_user_args])
+
+resource transid[int64]
+resource devid[int64]
+# This should be a resource but it's not returned by any syscall. What should we do about it?
+# resource qgroupid[int64]
+resource treeid[int64]
+resource dirid[int64]
+
+define BTRFS_PATH_MAX	BTRFS_PATH_NAME_MAX+1
+
+#This pads the struct to 1032 bytes. It was originally meant to pad to
+#1024 bytes, but when adding the flags field, the padding calculation
+#was not adjusted.
+define BTRFS_DEV_PAD	128-2-BTRFS_DEV_STAT_VALUES_MAX
+
+define BTRFS_VOL_MAX	BTRFS_VOL_NAME_MAX + 1
+
+#add a padding since fd in this struct should be 64 bits (is this the correct way?)
+btrfs_ioctl_vol_args {
+	fd	fd
+	dummy	const[0, int32]
+	name	array[int8, BTRFS_PATH_MAX]
+}
+
+btrfs_ioctl_vol_args_v2 {
+	fd	fd
+	transid	transid
+	flags	flags[btrfs_ioctl_vol_args_v2_flags, int64]
+	u1	btrfs_args_v2_u1
+	u2	btrfs_args_v2_u2
+}
+
+btrfs_args_v2_u1 [
+	inherit	btrfs_args_v2_u1_s1
+	unused	array[int64, 4]
+]
+
+btrfs_args_v2_u2 [
+	name		array[int8, BTRFS_PATH_MAX]
+	devid		devid
+	subvolid	int64
```

Comment:
> I've tried to go through everything and leave comments. I'm not a btrfs expert, though; this is the first time I've looked at these ioctls. And there are a lot of them.
> 
> Also, did you find documentation for these? I can't, so I just checked the kernel code...

Thank you so much Eric! I went through all the comments and fixed them. There are still a few places I'm not sure how to cope with, which I left comment under corresponding reviews. Another thing that I'm concerned with is that right now BTRFS_IOC_SET_RECEIVED_SUBVOL_32 and BTRFS_IOC_SEND_32 are commented out because they need MACROS which are defined in ioctl.c, and I feel like including a .c file shouldn't be the way to go. What do you think is the best way to deal with this?